include/picongpu/simulation/stage/SynchrotronRadiation.hpp

```diff
@@ -0,0 +1,252 @@
+            using namespace picongpu::particles::synchrotron::params;
+
+            //! from zq to FirstSynchrotronFunctionParams::logEnd
+            float_64 const log_start = std::log2(zq);
+            float_64 const log_step = (FirstSynchrotronFunctionParams::logEnd - log_start)
+                / (FirstSynchrotronFunctionParams::numberSamplePoints - 1);
+
+            float_64 integral = 0.;
+
+            float_64 xLeft;
+            float_64 xRight = zq;
+
+            for(unsigned int i = 0; i < FirstSynchrotronFunctionParams::numberSamplePoints - 1; ++i)
+            {
+                xLeft = xRight;
+                xRight = math::pow(2., log_start + log_step * (i + 1));
+                float_64 xMiddle = (xLeft + xRight) / 2.0;
+
+                //! try and catch errors in the bessel function
+                try
+                {
+                    float_64 yLeft = std::cyl_bessel_k(5.0 / 3.0, xLeft);
+                    float_64 yMiddle = std::cyl_bessel_k(5.0 / 3.0, xMiddle);
+                    /** computes the integral over one interval: [xLeft, xRight] using the
+                     *  exponential approximation between: [xLeft, xMiddle]
+                     */
+                    integral += integrateAsExponential(xLeft, xMiddle, xRight, yLeft, yMiddle);
+                }
+                catch(std::exception& e)
+                {
+                    std::cout << "Caught exception when precomputing firstSynchrotronFunction: " << e.what()
+                              << std::endl;
+                }
+            }
+            return zq * integral;
+        }
+
+        //! @note see paper: "Extended particle-in-cell schemes for physics in ultrastrong laser fields: Review and
+        //! developments" by A. Gonoskov et.Al.
+        static float_64 secondSynchrotronFunction(float_64 const x)
+        {
+            return x * std::cyl_bessel_k(2. / 3, x);
+        }
+
+    public:
+        /** Create a synchrotron radiation instance: precompute first and second synchrotron functions
+         *
+         * @param cellDescription mapping for kernels
+         */
+        SynchrotronRadiation(MappingDesc const cellDescription) : cellDescription(cellDescription)
+        {
+            //! for "InterpolationParams" struct
+            using namespace picongpu::particles::synchrotron::params;
+            auto data_space = DataSpace<2>{InterpolationParams::numberTableEntries, 2};
+            auto grid_layout = GridLayout<2>{data_space};
+
+            //! capture the space for table and variables
+            tableValuesF1F2 = std::make_shared<GridBuffer<float_X, 2>>(grid_layout);
+            failedRequirementQ = std::make_shared<GridBuffer<int32_t, 1>>(DataSpace<1>{1});
+            //! set the values of variables to false
+            failedRequirementQ->getHostBuffer().getDataBox()(DataSpace<1>{0}) = false;
+            failedRequirementPrinted = false;
```

Comment:
Why?

Reply:
made two new functions:
void setFailedRequirementQ(bool)
bool checkFailedRequirementQ()

Reply:
> Why?

Mainly [SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle) but an abstract principle by itself is, of course, not a good reason for doing things. Concretely, it reads better and abstracts away implementation details that I otherwise have to parse upon reading the code. When I read the function as a whole, I basically did the same abstraction in my head like "Okay, so this paragraph checks for failed requirements." but was only able to do that after I had deciphered what the construction of a shared pointer to a grid buffer uses its data space argument for and so on. That makes probably three or four of the [seven things I can hold in short-term memory](https://en.wikipedia.org/wiki/The_Magical_Number_Seven,_Plus_or_Minus_Two) which means it has flushed some other potentially important information out that I'll have to got and fetch later on again. That's an unnecessary burden to my brain.